### PR TITLE
feat: implement catalog scan segment progress

### LINKED ISSUE #771: Task: Catalog Scan Segment Progress
Planning document:
- `docs/tasks/000195-catalog-scan-segment-progress.md`
- Type: task
- ID: 000195
- Title: Task: Catalog Scan Segment Progress

Generated from document-first issue workflow.
Part of #767

## Summary

Implement RFC 0022 Phase 2 by making catalog checkpoint scan produce reusable
catalog-safe redo segment progress. The progress is in-memory only, owned by
`TransactionSystem`, and derived from durable facts: the current catalog
checkpoint replay boundary in `catalog.mtb`, the durable first-retained redo
marker from Phase 1, and sealed redo super-block metadata.

This task does not advance the durable redo retention marker and does not unlink
redo files. Later truncation planning must read the transaction-system progress
directly, or rebuild equivalent summaries from retained redo super-blocks when
the in-memory progress is absent.

## Context

RFC 0022 adds catalog-backed physical redo truncation in phases. This task is
only Phase 2, "Catalog Scan Segment Progress." Phase 1 already persisted
`first_redo_log_seq` in `catalog.mtb` and made startup recovery and catalog
checkpoint scan honor marker-aware retained-suffix discovery.

Issue Labels:
- type:task
- priority:medium
- codex

Parent RFC:
- `docs/rfcs/0022-catalog-backed-redo-log-truncation.md`

Target RFC Phase:
- Phase 2: Catalog Scan Segment Progress

Phase prerequisites relied on:
- Phase 1 marker/discovery contract exists.
- `Catalog::scan_checkpoint_batch` reads
  `snapshot.meta.first_redo_log_seq` and passes it into
  `discover_redo_log_files`.
- Redo segment super-blocks already expose sealed state and real redo CTS
  ranges through `RedoLogSegment::sealed_redo_range`.
- `Session::checkpoint_catalog` already exists, uses normal session admission,
  and requires an idle session.

Phase-local choices resolved:
- Do not introduce a public `CatalogCheckpointOutcome` in this phase.
  `Session::checkpoint_catalog(&mut self) -> Result<()>` remains the public
  producer operation.
- Store progress under `TransactionSystem`, not `Catalog`, so future
  `Session::truncate_redo_log` planning can consume the latest internal
  retention evidence directly without coupling caller order to a checkpoint
  return value.
- Store simple catalog-only progress:

```rust
pub(crate) struct CatalogRedoRetentionProgress {
    pub(crate) first_retained_file_seq: u32,
    pub(crate) catalog_replay_start_ts: TrxID,
    pub(crate) segments: Vec<CatalogSafeRedoSegment>,
}

pub(crate) struct CatalogSafeRedoSegment {
    pub(crate) file_seq: u32,
    pub(crate) redo_range: Option<RedoSegmentCtsRange>,
}

pub(crate) struct RedoSegmentCtsRange {
    pub(crate) min_cts: TrxID,
    pub(crate) max_cts: TrxID,
}
```

- `redo_range: None` means a sealed empty redo file. It is useful to retain
  because an empty sealed file can be a later whole-file truncation candidate,
  even though it has no CTS range.
- Do not add a generic `RedoRetentionProgress` wrapper or placeholder full
  retention state in this phase. Phase 3 can introduce a broader plan/result
  shape when table floors and dropped-table blockers are added.

Following-phase contract preserved and enabled:
- Phase 3 can read `TransactionSystem` catalog-safe segment progress directly.
- Phase 3 must still compute the global truncation floor from catalog progress,
  live-table replay bounds, and pending dropped-table replay bounds.
- Phase 3 must provide a fallback path when in-memory catalog progress is
  absent, such as immediately after restart. That fallback can rebuild equivalent
  segment summaries from retained sealed redo super-blocks plus durable
  `catalog_replay_start_ts`.
- No RFC phase-plan restructure is required. During task resolve, sync RFC 0022
  Phase 2 to record that the public checkpoint API stayed `Result<()>` and that
  catalog-safe scan progress is `TransactionSystem`-owned in-memory state.

Current behavior:
- `CatalogCheckpointBatch` records `replay_start_ts`, `safe_cts`, catalog row
  operations, included DDL count, and scan stop reason, but carries no redo
  segment progress.
- `Catalog::checkpoint_now` scans a batch and applies it, then wakes
  dropped-table purge on success. It returns `Result<()>`.
- `RedoReplayPlanner::plan_recovery` validates segment super-block metadata and
  selects replay segments, but sealed segment metadata is not returned as
  reusable catalog progress.
- `RedoLogStream` exposes accepted-prefix metadata for unsealed recovery tails,
  but Phase 2 must record sealed segment summaries because only sealed files can
  be future truncation candidates.

## Goals

- Expose sealed redo segment summaries from the redo scan/planner layer without
  changing redo replay semantics.
- Make catalog checkpoint scan collect retained sealed segment file sequences
  and CTS ranges while planning/scanning durable redo.
- Add a `TransactionSystem`-owned in-memory
  `CatalogRedoRetentionProgress` cache.
- Record catalog-safe segment progress only after the corresponding catalog
  checkpoint publish succeeds.
- Keep the progress conservative, monotonic, and restart-discardable.
- Keep `Session::checkpoint_catalog(&mut self) -> Result<()>` unchanged as the
  public API for this phase.
- Add focused tests for segment summary collection, progress update rules, and
  public checkpoint admission.

## Non-Goals

- Do not persist `CatalogRedoRetentionProgress`.
- Do not advance `first_redo_log_seq`.
- Do not add a production marker advancement API.
- Do not unlink, delete, truncate, or hole-punch redo files.
- Do not implement `Session::truncate_redo_log`.
- Do not compute global truncation floors.
- Do not retain dropped-table replay floors.
- Do not add table-floor blockers or truncation blocker taxonomy.
- Do not add background redo retention scheduling.
- Do not change redo super-block format or transaction redo payload format.
- Do not add a public `CatalogCheckpointOutcome` or require callers to pass a
  checkpoint outcome into truncation.

### PR DESCRIPTION
**Pull request by jiangzhe**
Closes #771 

## Summary by CodeRabbit

* **New Features**
  * Enhanced catalog checkpointing and recovery planning to compute and publish catalog-safe redo-retention progress, including sealed segment summaries for catalog scans.
* **Bug Fixes**
  * Record retention progress only when a checkpoint batch is successfully published; improved handling of no-op/canceled batches to avoid incorrect progress.
  * Strengthened validation around checkpoint replay to better reject invalid or inconsistent catalog state.
* **Tests**
  * Added/extended unit coverage for retention-progress derivation, catalog scan planning (sealed-empty/non-empty), checkpoint replay edge cases, and LWC/scan-row value append correctness.

### CODE REVIEW

File: doradb-storage/src/catalog/storage/mod.rs [hunk 1]
```diff
@@ -129,6 +149,801 @@ impl CatalogStorage {
...
+                    return Err(invalid_catalog_payload(format!(
+                        "single catalog row does not fit in LWC block: row_id={}",
+                        row.row_id
+                    )));
+                }
+            }
+            builder_end = row.row_id.saturating_add(1);
+        }
+
+        meta_pool.deallocate_page(temp_page);
```
[the remark this answers is not in the document]
Comment (jiangzhe) [reply doradb-storage/src/catalog/storage/mod.rs]: the append path is wrong: row => page => builder.

File: doradb-storage/src/catalog/storage/mod.rs [hunk 2]
```diff
@@ -139,6 +954,134 @@ pub(crate) struct CatalogDefinition {
...
+    if index_spec.cols.len() != key.vals.len() {
+        return false;
+    }
+    for (index_key, key_val) in index_spec.cols.iter().zip(&key.vals) {
+        let col_idx = index_key.col_no as usize;
+        if row.get(col_idx) != Some(key_val) {
+            return false;
+        }
+    }
+    true
```
[the remark this answers is not in the document]
Comment (jiangzhe) [reply doradb-storage/src/catalog/storage/mod.rs]: better solution is extracting the check outside of this function.
